### ISSUE TITLE
[Debug Tool] Detect unused variables

### ISSUE THREAD

**Issue by tfjmp** [End-to-end-provenance/dev#108]
We had a chat with @MKLau @NullCGT. Looking at the graph Connor generated: [ddg.txt](https://github.com/End-to-end-provenance/dev/files/1032417/ddg.txt) it appears that some variables created are not used to generate the end results. It would be relatively easy to detect such variables (they are "end" entities that are not "used" by any activity). Matt believe that could be useful, something akin to:

`warning: variable XXX was generated but not used`

**Comment by blernermhc** [comment]
I agree that if a user is using the script to run in its entirety like a program that unused data nodes are worth warning about.  I wonder, though, about users who use the script to document their work, and run it piecemeal in the console.  In that case, they might execute some lines and examine the values in RStudio's environment panel, using those computations as a sanity check of whether the computation looks like it is working.  

For example, a scientist might read some data from a csv file and then extract the data that falls between 2 dates.  The scientist might then call nrow to see how many rows of data are there, just to be sure that the data looks correct, not because they are going to use that in further calculations.

In the case of Connor's script, I am pretty certain that is why those unused variables exist.  He was doing an exercise to learn R that involved data exploration and answering questions like how many days there was precipitation.  That may not be representative of how a scientist would explore data but I wonder if we would see that kind of behavior, especially in console mode, where unused data should not be considered an error.  

Basically, the data might be consumed by the scientist's eyes even if the script does not consume them.  I just don't know how common that would be.

**Comment by MKLau** [comment]
Yeah, I will commonly add prints to my script as I'm working on it and then I'm usually lazy and don't go back and clean them up. 

I'll ponder this a bit more, but I can't think of a situation in which a case like this would produce essential provenance information.
